File: app/src/main/java/com/example/mz/udacitypopularmovies/TrailerAdapter.java
package com.example.mz.udacitypopularmovies;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.mz.udacitypopularmovies.data.TrailerEntry;
import com.example.mz.udacitypopularmovies.utilities.NetworkUtils;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.Arrays;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by mz on 2017-02-04.
 */
public class TrailerAdapter extends RecyclerView.Adapter<TrailerAdapter.ViewHolder> {

    private static final String LOG_TAG = TrailerAdapter.class.getSimpleName();

    private ArrayList<TrailerEntry> mTrailers;

    public TrailerAdapter() {
    }

    public void setTrailerData(TrailerEntry[] trailerData) {
        Log.i(LOG_TAG, "before setTrailerData there is " + getItemCount() + " elements in a view");
        if (trailerData == null) {
            mTrailers = null;
        } else if (mTrailers == null) {
            mTrailers = new ArrayList<>(Arrays.asList(trailerData));
        } else {
            mTrailers.addAll(Arrays.asList(trailerData));
        }
        notifyDataSetChanged();
        Log.i(LOG_TAG, "after setTrailerData there is " + getItemCount() + " elements in a view");
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        int layoutIdForItem = R.layout.trailer_details;
        LayoutInflater inflater = LayoutInflater.from(context);
        final boolean shouldAttachToParentImmediately = false;
        View view = inflater.inflate(layoutIdForItem, parent, shouldAttachToParentImmediately);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.bind(position);
    }

    @Override
    public int getItemCount() {
        if (mTrailers == null) {
            return 0;
        } else {
            return mTrailers.size();
        }
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final String LOG_TAG = ViewHolder.class.getSimpleName();
        @BindView(R.id.trailer_icon)
        ImageView iv_trailer_thumbnail;
        @BindView(R.id.trailer_name)
        TextView tv_trailer_name;

        public ViewHolder(View rootView) {
            super(rootView);
            ButterKnife.bind(this, rootView);
        }


        public void bind(int position) {
            if (mTrailers == null) {
                Log.d(LOG_TAG, "Called bind with null mReviews");
                return;
            }
            TrailerEntry trailer = mTrailers.get(position);

            Uri thumbnailLink = NetworkUtils.buildYoutubeThumbnail(trailer.key);
            Context context = super.itemView.getContext().getApplicationContext();
            Picasso.with(context).load(thumbnailLink).placeholder(R.mipmap.placeholder).into(iv_trailer_thumbnail);
            tv_trailer_name.setText(trailer.name);
        }

        @OnClick
        void openYoutubeTrailer() {
            int adapterPosition = getAdapterPosition();
            Log.e(LOG_TAG, "Clicked on position : " + adapterPosition);
            TrailerEntry trailer = mTrailers.get(adapterPosition);
            if (trailer == null) {
                return;
            }
            Uri youtubeMovie = NetworkUtils.buildYoutubeRequest(trailer.key);
            super.itemView.getContext().startActivity(new Intent(Intent.ACTION_VIEW, youtubeMovie));

        }
    }
}

File: app/src/main/java/com/mz/popmovies/TrailerAdapter.kt
package com.mz.popmovies

import com.mz.popmovies.utilities.NetworkUtils.buildYoutubeThumbnail
import com.mz.popmovies.utilities.NetworkUtils.buildYoutubeRequest
import androidx.recyclerview.widget.RecyclerView
import com.mz.popmovies.data.TrailerEntry
import android.view.ViewGroup
import android.view.LayoutInflater
import com.squareup.picasso.Picasso
import android.content.Intent
import android.util.Log
import android.view.View
import com.mz.popmovies.databinding.TrailerDetailsBinding
import java.util.*

/**
 * Created by mz on 2017-02-04.
 */
class TrailerAdapter : RecyclerView.Adapter<TrailerAdapter.ViewHolder>() {
    private var trailers: ArrayList<TrailerEntry> = ArrayList()
    private lateinit var binding: TrailerDetailsBinding
    fun setTrailerData(trailerData: List<TrailerEntry>) {
        Log.i(LOG_TAG, "before setTrailerData there is $itemCount elements in a view")
        trailers.addAll(trailerData)
        notifyDataSetChanged()
        Log.i(LOG_TAG, "after setTrailerData there is $itemCount elements in a view")
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        binding = TrailerDetailsBinding.inflate(inflater, parent, false)
        return ViewHolder(binding.root)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(position)
    }

    override fun getItemCount() = trailers.size

    inner class ViewHolder(rootView: View?) : RecyclerView.ViewHolder(rootView!!) {
        private val LOG_TAG = ViewHolder::class.java.simpleName

        fun bind(position: Int) {
            val trailer = trailers[position]
            val thumbnailLink = buildYoutubeThumbnail(trailer.key)
            Picasso.get().load(thumbnailLink).placeholder(R.mipmap.placeholder).into(binding.trailerIcon)
            binding.trailerName.text = trailer.name
            binding.root.setOnClickListener{
                openYoutubeTrailer()
            }
        }

        fun openYoutubeTrailer() {
            val adapterPosition = adapterPosition
            Log.e(LOG_TAG, "Clicked on position : $adapterPosition")
            val trailer = trailers[adapterPosition]
            val youtubeMovie = buildYoutubeRequest(trailer.key)
            val viewTrailerIntent = Intent(Intent.ACTION_VIEW, youtubeMovie)
            if (viewTrailerIntent.resolveActivity(super.itemView.context.packageManager) != null) {
                super.itemView.context.startActivity(viewTrailerIntent)
            }
            else {
                Log.w(LOG_TAG, "Youtube activity not resolved")
            }
        }
    }

    companion object {
        private val LOG_TAG = TrailerAdapter::class.java.simpleName
    }
}